### PR TITLE
Audit the installed headers for C++ compatibility

### LINKED ISSUE #13541: [5.2 & 5.3] Compiling C++ code using the OCaml C API results in name-mangled caml_state on Cygwin
Trying to compile opam (which uses C++ through mccs) on Cygwin with the 5.3 branch at 162a7f4869bb3136b48fd811408cc6f53396a0ea, fails with:
```
  ** Cannot resolve symbols for descriptor object:
   _ZTH10caml_state
  File "caml_startup", line 1:
  Error: Error during linking (exit code 2)
```
Whereas it compiles just fine with OCaml 4.14.2 with the same environment.

### PR DESCRIPTION
**Pull request by MisterDA**
C++ name mangling applies to symbols: variable and function names. The rule of thumb is to enclose public symbols declarations in blocks, to avoid C++ linkers look for C++ symbol names instead of C.

```c
#ifdef __cplusplus
extern "C" {
#endif

/* symbols go here */

#ifdef __cplusplus
}
#endif
```

Symbols protected by `CAML_INTERNALS` blocks need not to be covered. I've sometimes reordered a few functions to merge two sections protected by `CAML_INTERNALS`.

Headers that contain definitions incompatible with C++, such as `_Atomic`, also need protection.

There shouldn't be anything but comments before header guards to allow for [multiple include optimization](https://gcc.gnu.org/onlinedocs/cppinternals/Guard-Macros.html).

Fix #13541. Close #11557.

### CODE REVIEW

File: runtime/caml/osdeps.h [hunk 1]
```diff
@@ -139,12 +130,14 @@ CAMLextern int caml_win32_isatty(int fd);
 
 CAMLextern void caml_expand_command_line (int *, wchar_t ***);
 
-CAMLextern clock_t caml_win32_clock(void);
-
 CAMLextern value caml_win32_xdg_defaults(void);
 
 CAMLextern value caml_win32_get_temp_path(void);
 
+#include <time.h>
```
Comment (NickBarnes) [inline runtime/caml/osdeps.h]: I'd be inclined to move this to the top of the `_WIN32` section, for consistency.

**Review by NickBarnes — COMMENTED** [review]
Looks good to me; consistency is a good thing.

File: runtime/caml/osdeps.h (hunk 1 above)
Comment (MisterDA) [reply runtime/caml/osdeps.h]: Done, thanks.

**Comment by MisterDA** [comment]
I've added a test that checks [whether C++ stubs can be compiled and linked](https://github.com/ocaml/ocaml/pull/13591/commits/9ba790d284ff177a46f42ed0f2c9ed2e9b6305d8). I think it replaces and should close #11557. Thanks @kit-ty-kate for the tip!

This should be considered for 5.3 and 5.2.

**Comment by ghost** [comment]
Mangling of variable names only matters when namespaces are used, which is obviously not the case in OCaml. Therefore several of the added guards are unnecessary, e.g. `runtime/caml/address_class.h` doesn't need any.

**Comment by MisterDA** [comment]
> Therefore several of the added guards are unnecessary, e.g. `runtime/caml/address_class.h` doesn't need any.

Is that a strong blocker? Can we keep them for future use, or should I remove them?

**Comment by ghost** [comment]
No, it's not a blocker.

**Comment by Octachron** [comment]
@NickBarnes , did you have the time to finish your review? If possible I would like to include this fix in the next beta for 5.3 (that should happen next week hopefully).

**Comment by MisterDA** [comment]
I'm tempted to add the C++ guards to all headers, unconditionally, even encompassing `CAML_INTERNALS` sections…

**Review by NickBarnes — APPROVED** [review]
LGTM. Having a test is great. As noted in your comment, it would be nice if the test's `#include`s were auto-generated, but that's a problem for another time.

**Comment by NickBarnes** [comment]
@Octachron I apologise for the delay.

**Comment by NickBarnes** [comment]
Merged and cherry-picked to 5.3.

**Comment by Octachron** [comment]
I agree that cherry-picking 38b9c9e6a19b1c6bc57b6 to 5.3 is a good idea too to avoid diverging ocamltest tests.

**Comment by NickBarnes** [comment]
Done under @dra27's close supervision ....

**Pull request merged**